lib/ex_json_schema/validator.ex

```diff
@@ -8,19 +8,18 @@ defmodule ExJsonSchema.Validator do
   alias ExJsonSchema.Schema
   alias ExJsonSchema.Schema.Root
 
-  @type errors :: [{String.t, String.t}] | []
+  @type errors :: [%Error{}]
 
-  @spec validate(Root.t, ExJsonSchema.data) :: :ok | {:error, errors}
+  @spec validate(Root.t | ExJsonSchema.json, ExJsonSchema.data) :: :ok | {:error, errors} | no_return
```

Comment:
Is there a regular use case where this would not return (throw or raise)? If it only raises an error in an exceptional situation, I don't know if it should be added to the function spec. Otherwise pretty much every function would need that return type.

Reply:
I had to add the `no_return` to all the validation related functions because deep down we make a call to `raise_invalid_reference_error`, which raises an error. Dialyzer figures out (correctly) that no_return is also a possible return type and complains if it's missing.

Reply:
Hmm yeah. It can only happen when you pass an unresolved schema as a map and then it ends up validating that schema. It's not a recommended use case but it can happen, so essentially it is correct. Just leave it in then, if it makes Dialyzer happy :+1: